From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Charles Kerr <charles@charleskerr.com>
Date: Thu, 20 Jul 2023 14:18:19 -0500
Subject: fix: isURL() implementation

Modify Node's lib/internal/url.js isURL() to return the correct value for
URLs created both inside and outside of Node.

The isURL() impl in `main` is Electron-safe but can't be backported in
isolation because it relies on other changes from 18 to main. But we can
safely get there by trying the 18 version first (catching Node URLs),
then the `main` version (catching Electron URLs).

More background w/upstream links at
https://github.com/electron/electron/pull/39154#issuecomment-1644433388

This patch can be removed if upstreamed.

diff --git a/lib/internal/url.js b/lib/internal/url.js
index 36604c9de2179b5f6941dd9a7af66cdb03facd7f..27c3d1c1909ead49c4384d6d53d60c40115e9542 100644
--- a/lib/internal/url.js
+++ b/lib/internal/url.js
@@ -592,7 +592,9 @@ ObjectDefineProperties(URLSearchParams.prototype, {
  * @returns {self is URL}
  */
 function isURL(self) {
-  return self != null && ObjectPrototypeHasOwnProperty(self, context);
+  if (self != null && ObjectPrototypeHasOwnProperty(self, context))
+    return true; // a Node URL
+  return Boolean(self?.href && self.origin); // another impl's URL
 }
 
 class URL {
